data/account/participationRegistry.go

```diff
@@ -0,0 +1,385 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package account
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/util/db"
+)
+
+const maxBalLookback = 320
+
+// ParticipationID identifies a particular set of participation keys.
+type ParticipationID crypto.Digest
+
+// ParticipationRecord contains all metadata relating to a set of participation keys.
+type ParticipationRecord struct {
+	ParticipationID ParticipationID
+
+	Account     basics.Address
+	FirstValid  basics.Round
+	LastValid   basics.Round
+	KeyDilution uint64
+
+	LastVote               basics.Round
+	LastBlockProposal      basics.Round
+	LastCompactCertificate basics.Round
+	EffectiveFirst         basics.Round
+	EffectiveLast          basics.Round
+
+	// VRFSecrets
+	// OneTimeSignatureSecrets
+}
+
+// ParticipationAction is used when recording participation actions.
+type ParticipationAction int
+
+// ParticipationAction types
+const (
+	Vote ParticipationAction = iota
+	BlockProposal
+	CompactCertificate
+)
+
+var participationActionFields = map[ParticipationAction]string{
+	Vote:               "lastVoteRound",
+	BlockProposal:      "lastBlockProposalRound",
+	CompactCertificate: "lastCompactCertificateRound",
+}
+
+// ErrParticipationIDNotFound is used when attempting to update a set of keys which do not exist.
+var ErrParticipationIDNotFound = errors.New("the participation ID was not found")
+
+// ErrInvalidRegisterRange is used when attempting to register a participation key on a round that is out of range.
+var ErrInvalidRegisterRange = errors.New("key would not be active within range")
+
+// ErrUnknownParticipationAction is used when record is given something other than the known actions.
+var ErrUnknownParticipationAction = errors.New("unknown participation action")
+
+// ParticipationRegistry contain all functions for interacting with the Participation Registry.
+type ParticipationRegistry interface {
```

Comment:
don't forget to add a Close() method, which will close the underlying database.
( and also - please make sure to do that as well in all the unit tests )